mypy/semanal.py

```diff
@@ -972,11 +972,7 @@ def prepare_method_signature(self, func: FuncDef, info: TypeInfo, has_self_type:
             elif isinstance(functype, CallableType):
                 self_type = get_proper_type(functype.arg_types[0])
                 if isinstance(self_type, AnyType):
-                    if has_self_type:
-                        assert self.type is not None and self.type.self_type is not None
-                        leading_type: Type = self.type.self_type
-                    else:
-                        leading_type = fill_typevars(info)
+                    leading_type: Type = info.self_type if info.self_type else fill_typevars(info)
```

Comment:
TBH this is quite arbitrary. This will result in `Self` inferred if it is used explicitly somewhere else in the current class. Something like this will continue to error (unless I am missing something):
```python
from typing import Self

class C:
    x: Self
class D(C):
    def foo(self) -> None:
        self.x = self
```
But also as I mentioned in the issue, finding a heuristic that will work well without also causing big perf penalty is hard. Another option to consider is adding a flag for this (like `--force-implicit-self`) and then create and use self types _always_.

In general, I would not hurry with this, we need to carefully consider all options, since it may be hard to undo.

Reply:
> Something like this will continue to error (unless I am missing something):

I did not account for that, tweaked some code and using:
```py
self.setup_self_type()
leading_type: Type = info.self_type
```
to determine leading_type fixes it (`setup_self_type` wont replace the type if its already set)

*I am going to push this to the branch just to be able to run `mypy-primer` on it and see the results, its not final*

---

But something I noticed is that mypy seems to completely ignore inheritance when dealing with `Self`, something that `pyright` smartly deals with. Maybe the implementation of `Self` should be revisited and all these weird edge cases will disappear. I also dont personally like implementing a `--force-implicit-self` type, as this should be the default as stated in PEP 673. It should be sensible of mypy to properly deal and treat with these issues without having to enable extra flags

Reply:
> But something I noticed is that mypy seems to completely ignore inheritance when dealing with Self

What does this mean? Also you are talking so much about pyright, just go work with it.

Reply:
> What does this mean? Also you are talking so much about pyright, just go work with it.

I didnt mean to sound mean with the comment. Just wanted to point out that it might be better to deal with `Self` in a different way than `mypy` that might help to deal with all these edge cases that are being found as people start to use `Self` with `mypy`. I do not have any intent to turn mypy into pyright, was just pointing out a different implementation that might be better to deal with these cases.

I sincerely apologize if my comment came out as "pyright is better than mypy and mypy should follow its lead", it wasnt my intention. I am sorry